content/posts/faster-rfft-python/index.md

```diff
@@ -0,0 +1,276 @@
+
+for version, timing in timing_results.items():
+    if version == 'V0':
+        plt.axhline(y=1.0, color='r', linestyle='--', label='baseline')  # baseline
+    else: 
+        plt.plot(
+            np.arange(p_min, p_max + 1),
+            timing_results['V0'] / timing,
+            label=version,
+            marker='o'
+        )
+
+plt.xticks(
+    ticks=np.arange(p_min, p_max + 1),
+    labels=np.arange(p_min, p_max + 1),
+)
+plt.xlabel('log2(n)')
+plt.ylabel('Speed-up Factor')
+plt.grid()
+plt.legend()
+plt.show()
+```
+
+The above code sets up a framework to compare different RFFT implementations. It defines a dictionary `rfft_callers` to hold different versions of RFFT caller functions. It then computes the timing for each version and plots the speed-up factor compared to the baseline version (V0), which is provided below. 
+
+We are now ready to explore different implementations of RFFT using pyFFTW and see how we can improve the performance! Let's start with a baseline implementation of RFFT using pyFFTW.
+
+### Baseline (V0)
+```python
+import numpy as np
+import pyfftw
+
+
+def rfft_caller_v0(T):
+    """
+    Returns the RFFT of a real-valued array T
+    """
+    rfft_obj = pyfftw.builders.rfft(np.empty(len(T)))
+    
+    return rfft_obj(T)
+```
+
+### First Attempt: Byte-align the input array in advance (V1)
+`pyfftw` internally byte-aligns the input array by default to provide better performance. One potential optimization is to create a byte-aligned array in advance, and reuse this array. This allows `pyfftw` to avoid unnecessary copies of the input data that might happen within the RFFT computation.  We can also allow `pyfftw` to overwrite our array as it should not affect our original data. Here is the modified version:
+
+```python
+class rfft_caller_v1:
+    def __init__(self):
+        self.real_arr = None
+
+    def __call__(self, T):
+        if self.real_arr is None or len(T) != len(self.real_arr):
+            self.real_arr = pyfftw.empty_aligned(len(T), dtype='float64')
+        
+        rfft_obj = pyfftw.builders.rfft(self.real_arr, overwrite_input=True, avoid_copy=True)
+        self.real_arr[:] = T
+        rfft_obj.execute()
+        
+        return rfft_obj.output_array
+```
+
+In this modified version, an instance of the class `rfft_caller_v1` can be used to compute the RFFT. It holds the byte-aligned array `real_arr`, as an attribute, after the first call. When the `__call__` method is invoked again, it checks if the size of the input array `T` matches the size of the existing aligned array, and if yes, it reuses the existing byte-aligned array. Also, note that we intentionally do not save the RFFT object as attribute and do not use it across multiple calls in this version. We will explore that in a later version. Let's update the dictionary `rfft_callers` to include this new version:
+
+```python
+rfft_callers = {
+    'V0': rfft_caller_v0,
+    'V1': rfft_caller_v1(),
```

Comment:
It's a little weird to see `rfft_caller_v0` and then `rfft_caller_v1()` with the additional parentheses.